src/charm.py

```diff
@@ -43,6 +44,77 @@ def __init__(self, *args: typing.Any) -> None:
+        self.database: typing.Optional[DatabaseRequires] = self._setup_database_requirer(
+            "mysql", "spring-boot"
+        )
+
+    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
+        """Set up a DatabaseRequires instance.
+
+        The DatabaseRequires instance is an interface between the charm and various data providers.
+        It handles those relations and emit events to help us abstract these integrations.
+
+        Args:
+            relation_name: Name of the data relation
+            database_name: Name of the database (can be overwritten by the provider)
+
+        Returns:
+            DatabaseRequires object
+        """
+        database_requirer = DatabaseRequires(
+            self,
+            relation_name=relation_name,
+            database_name=database_name,
+        )
+        self.framework.observe(database_requirer.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[relation_name].relation_broken, self._on_relation_broken)
+        return database_requirer
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application.
+
+        Args:
+            event: The DatabaseCreatedEvent object
+        """
+        logger.debug("Database credentials are received: %s", event.username)
+        self.reconciliation(event)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation broken event.
+
+        Args:
+            event: The RelationBrokenEvent object
+        """
+        logger.debug("Relation with database removed: %s", event.relation.name)
+        self.reconciliation(event)
+
+    def _datasource(self) -> dict[str, str]:
+        """Compute datasource dict and return it.
+
+        Returns:
+            Dict containing details about the data provider integration
+        """
+        if self.database:
+            relations_data = list(self.database.fetch_relation_data().values())
+            if relations_data:
+                # There can be only one database integrated at a time
+                # cf: metadata.yaml
+                data = relations_data[0]
+                if "endpoints" in data and "username" in data and "password" in data:
+                    # We assume that the relation follows the following json schema:
+                    # https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/mysql_client/v0/schemas/provider.json
+                    database_name = data.get("database", self.database.database)
+                    endpoint = data["endpoints"].split(",")[0]
+                    return {
+                        "username": data["username"],
+                        "password": data["password"],
+                        "url": f"jdbc:mysql://{endpoint}/{database_name}",
+                    }
+        return {
```

Comment:
What is the reason for returning this? What about raises an error? It doesn't seem like this is valid data that could be used to connect to a database

Reply:
This seem to not be in the written specification (@gregory-schiano) but we are supposed to overwrite the datasource details each time to make sure that they can't be decided by the application.

Reply:
Are there any alternatives? This seems to be something like returning an empty value in response to some data missing on the input. If the data we need is missing - isn't that an error? Maybe I'm misunderstanding the purpose of the function

Reply:
It's explained in the specs, we don't want to allow the user to specify its own database parameters in the application-config, therefore we will empty these fields by default to override user entries.
Concerning the "missing" data, Spring Boot can work "in degraded" when the connection to the database isn't working (readonly and using its cache), so I don't want to block the application from starting because of an integration problem. We're talking about an edge case here.

So as per the specs, it must be implemented this way, there's nothing wrong there